R/utils.R

```diff
@@ -134,14 +134,40 @@ check_modules_datanames <- function(modules, datanames) {
 }
 
 #' @rdname check_modules_datanames
-check_modules_datanames_html <- function(modules,
-                                         datanames) {
+check_reserved_datanames <- function(datanames) {
+  reserved_datanames <- datanames[datanames %in% c("all", ".raw_data")]
+  if (length(reserved_datanames) == 0L) {
+    return(NULL)
+  }
+
+  tags$span(
+    to_html_code_list(reserved_datanames),
+    sprintf(
+      "%s reserved for internal use. Please avoid using %s as %s.",
+      pluralize(reserved_datanames, "is", "are"),
+      pluralize(reserved_datanames, "it", "them"),
+      pluralize(reserved_datanames, "a dataset name", "dataset names")
+    )
+  )
+}
+
+#' @rdname check_modules_datanames
+check_modules_datanames_html <- function(modules, datanames) {
   check_datanames <- check_modules_datanames_recursive(modules, datanames)
   show_module_info <- inherits(modules, "teal_modules") # used in two contexts - module and app
+
+  reserved_datanames <- check_reserved_datanames(datanames)
+
   if (!length(check_datanames)) {
-    return(TRUE)
+    out <- if (is.null(reserved_datanames)) {
+      TRUE
+    } else {
+      reserved_datanames
+    }
+    return(out)
```

Comment:
I'm not too familiar with this code, but returning early omits the checks on missing_datanames below

Reply:
This returns early when there are no messages on missing datanames.

The actual check happens earlier

Reply:
Sorry, I checked more carefully the code and now I think I understand it. Although I find it more readable to use && instead of assignment on if clauses that have else statements, or simply using several times return inside the conditions.

Reply:
I think here is a matter of style. I don't like any of the possible styles so I ended up with a single return.

I'm flexible on this, is the suggestion below ok, or would you prefer an alternative implementation?

### 1. `if` ... `else` ...

```r
  if (!length(check_datanames) && is.null(reserved_datanames)) {
    return(TRUE)
  } else if (!length(check_datanames)) {
    return(shiny::tagList(reserved_datanames))
  }
```

<details>

<summary>Other options</summary>

### 2. 2 separate early returns

```r
  if (!length(check_datanames) && is.null(reserved_datanames)) {
    return(TRUE)
  } 
  if (!length(check_datanames)) {
    return(shiny::tagList(reserved_datanames))
  }
```

### 3. As is but double return

```r
  if (!length(check_datanames)) {
    if (is.null(reserved_datanames)) {
      return(TRUE)
    } else {
      return(shiny::tagList(reserved_datanames))
    }
  }
```

### 4. As is

```r
  if (!length(check_datanames)) {
    out <- if (is.null(reserved_datanames)) {
      TRUE
    } else {
      shiny::tagList(reserved_datanames)
    }
    return(out)
  }
```

As sidenote my favourite in terms of readability would be: (if we were using `%||%`)


```r
  if (!length(check_datanames)) {
    reserved_datanames %||% return(TRUE)
    return(shiny::tagList(reserved_datanames))
  }
```

</details>